sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/RowConverter.scala

```diff
@@ -0,0 +1,172 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+package com.azure.cosmos.spark
+
+import java.sql.{Date, Timestamp}
+import java.util
+
+import com.azure.cosmos
+import com.azure.cosmos.spark.CosmosLoggingTrait
+import com.fasterxml.jackson.databind.{JsonNode, ObjectMapper}
+import com.fasterxml.jackson.databind.node.{ArrayNode, NullNode, ObjectNode}
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.Row
+import org.apache.spark.sql.catalyst.expressions.GenericRowWithSchema
+import org.apache.spark.sql.catalyst.expressions.UnsafeMapData
+import org.apache.spark.sql.types.{BinaryType, BooleanType, DataType, DateType, DecimalType, DoubleType, FloatType, IntegerType, LongType, _}
+//import org.json.{JSONArray, JSONObject}
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.util.ArrayData
+import org.apache.spark.unsafe.types.UTF8String
+
+import scala.collection.JavaConverters._
+import scala.collection.immutable.HashMap
+import scala.collection.mutable.ListBuffer
+
+/**
+ * TODO add more unit tests for this class to CosmosRowConverterSpec.
+ */
+
+object CosmosRowConverter
+  extends Serializable
+    with CosmosLoggingTrait {
+
+  // TODO moderakh make this configurable
+  val objectMapper = new ObjectMapper();
+
+  def rowToObjectNode(row: Row): ObjectNode = {
+
+    val jsonObject: ObjectNode = objectMapper.createObjectNode();
+    row.schema.fields.zipWithIndex.foreach({
+      case (field, i) => {
+        val jsonValue = convertToJson(row.get(i), field.dataType, isInternalRow = false)
+        addJsonPrimitive(jsonValue, field.name, jsonObject)
+      }
+    })
+    jsonObject
+  }
+
+  def internalRowToObjectNode(internalRow: InternalRow, schema: StructType): ObjectNode = {
+    val jsonObject: ObjectNode = objectMapper.createObjectNode();
+    schema.fields.zipWithIndex.foreach({
+      case (field, i) => {
+        val jsonValue = convertToJson(internalRow.get(i, field.dataType), field.dataType, isInternalRow = true)
+        addJsonPrimitive(jsonValue, field.name, jsonObject)
+      }
+    })
+    jsonObject
+  }
+
+  private def addJsonPrimitive(jsonValue: Any, fieldName: String, objectNode : ObjectNode) : Unit = {
+    jsonValue match {
+      case element: Boolean => objectNode.put(fieldName, element.asInstanceOf[Boolean])
+      case element: String => objectNode.put(fieldName, element.asInstanceOf[String])
+      case element: Double => objectNode.put(fieldName, element.asInstanceOf[Double])
+      case element: Float => objectNode.put(fieldName, element.asInstanceOf[Float])
+      case element: Long => objectNode.put(fieldName, element.asInstanceOf[Long])
+      case element: Int => objectNode.put(fieldName, element.asInstanceOf[Int])
+      case element: JsonNode => objectNode.set(fieldName, element.asInstanceOf[JsonNode])
+      case _ => objectNode.putNull(fieldName)
+    }
+  }
+
```

Comment:
Where is this implementation coming from OLAP, built-in Spark connectors like CSVDataDSource? I though Spark also added capability to transform DataFrame forma nd to json - my gut feeling is that it would be good to stick with taht one.

Reply:
oh please don't review RowConverter yet. this class is evolving ...

Reply:
Row -> ObjectNode, I didn't find any out of the box suitable Row -> ObjectNode conversion.

based on my reading, a few interesting things which I found:
1) There is a "private internal" class converter in the spark code 
https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonGenerator.scala
As this is a private class not suitable for us.

2) The other option is serializing Row to a json String and passing Json string to CosmosClient. Not a good options as it requires parsing the string to extract PK (perf overhead)

3) the other option is dealing with 
https://spark.apache.org/docs/3.0.1/api/java/org/apache/spark/sql/Row.html#jsonValue--
which again doesn't seem to be a good option. this on the Scala code seem to be a private method not a public one.

4) one other thing I would like to read about is Row Encoders and Decoders:
https://jaceklaskowski.gitbooks.io/mastering-spark-sql/content/spark-sql-Encoder.html

The RowConverter code in this PR is a rewritten version of Row -> json that we have in the existing OLTP spark connector.
I rewrote it to work with jackson with some fixes and unit tests.

Are you referring to something else?

Reply:
I added this to TODO section. This requires more discussion/investigation. Will be done in a separate PR.